### DISCUSSION TITLE
大家用过分布式编译功能吗？

### DISCUSSION THREAD

**Discussion by Domain** [xmake-io/xmake#4383] [General]
很少看见有人讨论这个功能，issue也没什么人提，是很少人用还是大家都没遇到问题？我一直很关注这个功能，但是我自己使用问题很多，几乎没法使用。
我遇到的问题包括
- 卡死。所有参与编译的机器都没反应，三个超时都配置了，但是就是卡住没反应。
- 全部变成distcc了，也就是本地不参与编译了，导致编译更慢。
还有其他一些问题，记不清楚了，反正就是几乎没法使用，或者比单机编译更慢。

**Comment by waruqi** [comment]
目前这个使用不多，可以单个问题 提供复现过程和完整 logs ，或者直接自己调试下源码   目前我这环境暂时无法复现这些问题

**Reply by Domain** [reply]
我倒是想调试，但是这需要先看懂这些代码才行，目前没这个时间。现在遇到最多的就是第二个问题，也就是所有文件都变成distcc了，只有一开始有几个文件本地编译，之后就再也没有了，就像是本地编译的几个文件永远无法完成。

**Comment by Domain** [comment]
目前发现，卡死可能是因为远程编译的xmake进程异常退出了，应该是所有远程的都出了问题，然后本地xmake并不会自动转为本地编译，一直卡着，我所有超时时间都设置为5秒。异常退出的xmake没有coredump，日志只有大量session connect和session end，有部分有一个出错信息：error: too large data size(84511516)。
我觉得联合编译的调度很有问题，就是除了一开始，后面都看不到本地编译，经常出现等待很久，甚至卡死。超时机制似乎也有点问题。
我希望输出时，显示这个文件是在哪台机器编译的，详细信息时，能够显示每台机器的负载情况，就是发送了多少文件过去，收到了多少文件，能够更清晰看到卡在哪台机器上。

**Reply by waruqi** [reply]
这块等后面有空了 我再整体优化下

**Comment by Domain** [comment]
我发现，即使xmake进程没有异常退出，gcc进程异常退出了，例如被kill掉，分布式编译也会卡死。
而且现在好像是需要等待所有远程编译返回之后才会链接，就算其他exe都已经编译完了，只要还有一个exe的cpp还在远程编译，其他exe也不会链接。
还有，参与分布式编译的机器配置不一定一样，这就导致编译有快有慢，会出现所有快的机器已经编译好了，都在等待慢的机器。我觉得当所有任务分配完之后，有机器空闲了，就应该把正在运行的任务再分配出去，哪个返回快就用哪个结果，最大限度利用。

**Reply by Domain** [reply]
更正一下，gcc被kill之后，xmake没有卡死，只是因为我用了unity编译，编译时间比较长，而xmake也没有重新分配任务的提示，所以以为卡死了

**Reply by waruqi** [reply]
好的，回头我看下，最近比较忙

**Comment by Domain** [comment]
感觉只要有空闲机器，正在运行的任务能够重新调度到这台机器上，就能很大限度解决卡死问题，因为最差结果就是全部分配到本地

**Comment by yushroom** [comment]
我试用下有几个问题：
1. 开distcc， server大部分情况下都跑不满
2. client新加的toolchain，server上不认识，必须在server的xmake的toolchain目录下新建才行
